src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -308,45 +307,38 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
 
                         foreach (KeyValueWatcher changeWatcher in cacheExpiredWatchers.Concat(cacheExpiredMultiKeyWatchers))
                         {
                             UpdateCacheExpirationTime(changeWatcher);
                         }
 
-                        foreach (KeyValuePair<KeyValueIdentifier, KeyValueChange> kvp in keyValueChanges)
-                        {
-                            KeyValueChange keyValueChange = kvp.Value;
-
-                            if (keyValueChange.ChangeType == KeyValueChangeType.Modified)
-                            {
-                                applicationSettings[keyValueChange.Key] = keyValueChange.Current;
-                                watchedSettings[kvp.Key] = keyValueChange.Current;
-                            }
-                            else if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)
-                            {
-                                applicationSettings.Remove(keyValueChange.Key);
-                                watchedSettings.Remove(kvp.Key);
-                            }
-
-                            // Invalidate the cached Key Vault secret (if any) for this ConfigurationSetting
-                            foreach (IKeyValueAdapter adapter in _options.Adapters)
-                            {
-                                adapter.InvalidateCache(keyValueChange.Current);
-                            }
-                        }
-
-                        foreach (KeyValueChange change in changedKeyValuesCollection)
+                        foreach (KeyValueChange change in keyValueChanges.Concat(changedKeyValuesCollection))
                         {
-                            if (change.ChangeType == KeyValueChangeType.Deleted)
+                            if (change.ChangeType == KeyValueChangeType.Modified)
                             {
-                                applicationSettings.Remove(change.Key);
+                                ConfigurationSetting setting = change.Current;
+                                ConfigurationSetting settingCopy = new ConfigurationSetting(setting.Key, setting.Value, setting.Label, setting.ETag);
+                                watchedSettings[new KeyValueIdentifier(change.Key, change.Label)] = settingCopy;
+                                foreach (Func<ConfigurationSetting, ValueTask<ConfigurationSetting>> func in _options.Mappers)
+                                {
+                                    setting = await func(setting).ConfigureAwait(false);
+                                }
+                                if (setting == null)
+                                {
+                                    mappedData.Remove(change.Key);
+                                }
+                                else
+                                {
+                                    mappedData[change.Key] = setting;

```

Comment:
Shouldn't that be `setting.Key`?

If so, raises an interesting question. How do we know what to remove if they modify the key.